startup.s

```diff
@@ -0,0 +1,43 @@
+	.section .exception
+	.globl _start
+	.type _start, function
+
+/* Boot code at 0xbfc00000 */
+_start:	la	$sp, _estack	# Set stack pointer.
+	la	$gp, _gp	# Prepare global pointer.
+
+	/* Copy .data from ROM to RAM.
+	 * .data is located in ROM just after .text, so it starts at _etext.
+	 * It should land at the beggining of RAM, which is pointed at by _data.
+	 * Copy words one by one, until _edata is reached. */
+	la	$t1, _data	# dest = _data
+	la	$t2, _edata	# dest_end = _edata
+	la	$t3, _etext	# src = _etext
+copy_next_word:
+	lw	$t0, 0($t3)	# x = *src
+	sw	$t0, 0($t1)	# *dest = x
+	addiu	$t1, 4		# dest++
+	addiu	$t3, 4		# src++
+	bne	$t1, $t2, copy_next_word
+	  nop			# jump if dest != dest_end
+
+	/*     Clean .bss. It starts at _bss and ends at _ebss. */
+	la	$t1, _bss	# dest = _bss
+	la	$t2, _ebss	# dest_end = _ebss
+clear_next_word:
+	/*     Everybody loves MIPS $zero register. */
+	sw	$zero, 0($t1)	# *dest = 0
+	addiu	$t1, 4		# dest++
```

Comment:
Apparently I should not have to insert instructions into the delay slot at all. The assembler is aware of them, and automatically placed `addiu` right after `bne`:

    Dump of assembler code for function copy_next_word:
    0xbfc00028 <+0>:	lw	t0,0(t3)
    0xbfc0002c <+4>:	sw	t0,0(t1)
    0xbfc00030 <+8>:	addiu	t1,t1,4
    0xbfc00034 <+12>:	bne	t1,t2,0xbfc00028 <copy_next_word>
    0xbfc00038 <+16>:	addiu	t3,t3,4
    0xbfc0003c <+20>:	nop

Similarly, for `jal`:

    0xbfc00060 <+16>:	jal	0xbfc000a8 <kernel_main>
    0xbfc00064 <+20>:	nop

It does not swap instructions blindly if it could modify behaviour, and places `nop`s when needed:

    0xbfc00054 <+4>:	addiu	t1,t1,4
    0xbfc00058 <+8>:	bne	t1,t2,0xbfc00050 <clear_next_word>
    0xbfc0005c <+12>:	nop

Therefore we can simply ignore all delay slots. We can also choose to manage them on our own by disabling this feature with `.set noreorder`.